src/azul/maintenance.py

```diff
@@ -0,0 +1,355 @@
+                       actual_start=self._format_datetime(self.actual_start),
+                       actual_end=self._format_datetime(self.actual_end))
+        return result
+
+    @classmethod
+    def _parse_datetime(cls, value: str | None) -> datetime | None:
+        return None if value is None else parse_dcp2_datetime(value)
+
+    @classmethod
+    def _format_datetime(cls, value: datetime | None) -> str | None:
+        return None if value is None else format_dcp2_datetime(value)
+
+    @property
+    def start(self):
+        return self.actual_start or self.planned_start
+
+    @property
+    def end(self):
+        return self.actual_end or self.start + self.planned_duration
+
+    def validate(self):
+        require(isinstance(self.planned_start, datetime),
+                'No planned start', self.planned_start)
+        require(self.planned_start.tzinfo == UTC)
+        require(isinstance(self.description, str) and self.description,
+                'Invalid description', self.description)
+        for impact in self.impacts:
+            impact.validate()
+        reject(self.actual_end is not None and self.actual_start is None,
+               'Event has end but no start')
+        require(self.start < self.end,
+                'Event start is not before end')
+
+
+@attrs.define
+class MaintenanceSchedule:
+    events: list[MaintenanceEvent]
+
+    @classmethod
+    def from_json(cls, schedule: JSON):
+        return cls(events=list(map(MaintenanceEvent.from_json, schedule['events'])))
+
+    def to_json(self) -> JSON:
+        return dict(events=[e.to_json() for e in self.events])
+
+    def validate(self):
+        for e in self.events:
+            e.validate()
+        starts = set(e.start for e in self.events)
+        require(len(starts) == len(self.events),
+                'Start times are not distinct')
+        # Since starts are distinct, we'll never need the end as a tie-breaker
+        intervals = [(e.start, e.end) for e in self.events]
+        require(intervals == sorted(intervals),
+                'Events are not sorted by start time')
+        values = list(flatten(intervals))
+        require(values == sorted(values),
+                'Events overlap', values)
+        reject(len(self._active_events()) > 1,
+               'More than one active event')
+        require(all(e.actual_start is None for e in self.pending_events()),
+                'Active event mixed in with pending ones')
+
+    def pending_events(self) -> list[MaintenanceEvent]:
+        """
+        Returns a list of pending events in this schedule. The elements in the
+        returned list can be modified until another method is invoked on this
+        schedule. The modifications will be reflected in ``self.events`` but the
+        caller is responsible for ensuring they don't invalidate this schedule.
+        """
+        events = enumerate(self.events)
+        for i, e in events:
```

Comment:
This looks peculiar. Is there a reason for not writing
```suggestion
        for i, e in enumerate(self.events):
```
?